89944f6dfa0e3aeabef8abfa10ecd2649be30c49 from upstream, modified to not
conflict on src/main.cpp

From 15200f203a2ff62b7c0bb1c6849e874f545597cc Mon Sep 17 00:00:00 2001
From: Bart Ribbers <bribbers@disroot.org>
Date: Thu, 16 Mar 2023 09:39:50 +0100
Subject: [PATCH] Add option to blur background

---
 CMakeLists.txt                        | 17 +++++++++++++++--
 src/CMakeLists.txt                    | 11 ++++++++++-
 src/contents/ui/SettingsComponent.qml | 16 +++++++++++++++-
 src/contents/ui/SettingsDialog.qml    |  3 +--
 src/contents/ui/main.qml              |  6 ++++++
 src/terminalsettings.kcfg             |  3 +++
 src/util.cpp                          | 19 +++++++++++++++++++
 src/util.h                            |  2 ++
 8 files changed, 71 insertions(+), 6 deletions(-)

diff --git a/CMakeLists.txt b/CMakeLists.txt
index f6b141c..0b13e3c 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -38,8 +38,21 @@ ecm_setup_version(${PROJECT_VERSION}
 )
 ################# Find dependencies #################
 
-find_package(Qt5 ${QT_MIN_VERSION} REQUIRED NO_MODULE COMPONENTS Core Quick Test Gui Svg QuickControls2)
-find_package(KF5 ${KF5_MIN_VERSION} REQUIRED COMPONENTS Kirigami2 I18n Config CoreAddons)
+find_package(Qt5 ${QT_MIN_VERSION} REQUIRED NO_MODULE COMPONENTS 
+    Core 
+    Quick 
+    Test 
+    Gui 
+    Svg 
+    QuickControls2
+)
+find_package(KF5 ${KF5_MIN_VERSION} REQUIRED COMPONENTS 
+    Kirigami2 
+    I18n 
+    Config 
+    CoreAddons 
+    WindowSystem
+)
 
 ecm_find_qmlmodule(org.kde.kirigamiaddons.labs.mobileform 0.1)
 ecm_find_qmlmodule(QMLTermWidget 1.0)
diff --git a/src/CMakeLists.txt b/src/CMakeLists.txt
index 36d543a..1236bed 100644
--- a/src/CMakeLists.txt
+++ b/src/CMakeLists.txt
@@ -16,6 +16,15 @@ add_executable(qmlkonsole ${qmlkonsole_SRCS} ${RESOURCES})
 kconfig_add_kcfg_files(qmlkonsole GENERATE_MOC terminalsettings.kcfgc)
 target_include_directories(qmlkonsole PRIVATE ${CMAKE_CURRENT_BINARY_DIR} ${CMAKE_BINARY_DIR})
 
-target_link_libraries(qmlkonsole Qt5::Core Qt5::Qml Qt5::Quick Qt5::Svg KF5::I18n KF5::ConfigGui KF5::CoreAddons)
+target_link_libraries(qmlkonsole 
+    Qt5::Core 
+    Qt5::Qml 
+    Qt5::Quick 
+    Qt5::Svg 
+    KF5::I18n 
+    KF5::ConfigGui 
+    KF5::CoreAddons 
+    KF5::WindowSystem
+)
 install(TARGETS qmlkonsole ${KDE_INSTALL_TARGETS_DEFAULT_ARGS})
 install(FILES terminalsettings.kcfg DESTINATION ${KCFG_INSTALL_DIR})
diff --git a/src/contents/ui/SettingsComponent.qml b/src/contents/ui/SettingsComponent.qml
index 6072ee1..8c3ce6e 100644
--- a/src/contents/ui/SettingsComponent.qml
+++ b/src/contents/ui/SettingsComponent.qml
@@ -237,7 +237,7 @@ ColumnLayout {
                 }
             }
             
-            MobileForm.FormDelegateSeparator { below: fontFamilyDelegate }
+            MobileForm.FormDelegateSeparator { below: opacityDelegate }
             
             MobileForm.FormComboBoxDelegate {
                 id: opacityDelegate
@@ -286,6 +286,20 @@ ColumnLayout {
                     }
                 }
             }
+            
+            MobileForm.FormDelegateSeparator { above: opacityDelegate; below: blurDelegate }
+            
+            MobileForm.FormSwitchDelegate {
+                id: blurDelegate
+                text: i18n("Blur Background")
+                checked: TerminalSettings.blurWindow
+                
+                onCheckedChanged: {
+                    TerminalSettings.blurWindow = checked;
+                    TerminalSettings.save();
+                    Util.setBlur(applicationWindow().pageStack, TerminalSettings.blurWindow);
+                }
+            }
         }
     }
     
diff --git a/src/contents/ui/SettingsDialog.qml b/src/contents/ui/SettingsDialog.qml
index 1d11506..3602aa8 100644
--- a/src/contents/ui/SettingsDialog.qml
+++ b/src/contents/ui/SettingsDialog.qml
@@ -21,11 +21,10 @@ Kirigami.Dialog {
     preferredWidth: Kirigami.Units.gridUnit * 35
     
     Kirigami.Theme.inherit: false
-        Kirigami.Theme.colorSet: Kirigami.Theme.Window
+    Kirigami.Theme.colorSet: Kirigami.Theme.Window
     
     Control {
         id: control
-        implicitHeight: Kirigami.Units.gridUnit * 20
         
         leftPadding: 0
         rightPadding: 0
diff --git a/src/contents/ui/main.qml b/src/contents/ui/main.qml
index 37e620f..3394cc9 100644
--- a/src/contents/ui/main.qml
+++ b/src/contents/ui/main.qml
@@ -26,6 +26,12 @@ Kirigami.ApplicationWindow {
     
     color: "transparent"
     
+    Component.onCompleted: {
+        if (TerminalSettings.blurWindow) {
+            Util.setBlur(pageStack, true);
+        }
+    }
+    
     // pop pages when not in use
     Connections {
         target: applicationWindow().pageStack
diff --git a/src/terminalsettings.kcfg b/src/terminalsettings.kcfg
index 4214e75..2a0af00 100644
--- a/src/terminalsettings.kcfg
+++ b/src/terminalsettings.kcfg
@@ -25,5 +25,8 @@
         <entry key="actions" type="StringList">
             <default></default>
         </entry>
+        <entry key="blurWindow" type="bool">
+            <default>false</default>
+        </entry>
     </group>
 </kcfg>
diff --git a/src/util.cpp b/src/util.cpp
index 0797b3c..623c5b2 100644
--- a/src/util.cpp
+++ b/src/util.cpp
@@ -3,6 +3,10 @@
 
 #include "util.h"
 
+#include <QQuickWindow>
+
+#include <KWindowEffects>
+
 Util::Util(QObject *parent) 
     : QObject{ parent }
 {}
@@ -12,3 +16,18 @@ uint Util::getKeyFromString(QString key)
     QKeySequence seq = QKeySequence(key);
     return seq.count() > 0 ? seq[0] : 0;
 }
+
+void Util::setBlur(QQuickItem *item, bool blur)
+{
+    auto setWindows = [item, blur]() {
+        auto reg = QRect(QPoint(0, 0), item->window()->size());
+        KWindowEffects::enableBackgroundContrast(item->window(), blur, 1, 1, 1, reg);
+        KWindowEffects::enableBlurBehind(item->window(), blur, reg);
+    };
+
+    disconnect(item->window(), &QQuickWindow::heightChanged, this, nullptr);
+    disconnect(item->window(), &QQuickWindow::widthChanged, this, nullptr);
+    connect(item->window(), &QQuickWindow::heightChanged, this, setWindows);
+    connect(item->window(), &QQuickWindow::widthChanged, this, setWindows);
+    setWindows();
+}
diff --git a/src/util.h b/src/util.h
index 7f19b34..030704b 100644
--- a/src/util.h
+++ b/src/util.h
@@ -5,6 +5,7 @@
 
 #include <QObject>
 #include <QKeySequence>
+#include <QQuickItem>
 
 class Util : public QObject
 {
@@ -20,5 +21,6 @@ public:
     }
     
     Q_INVOKABLE uint getKeyFromString(QString key);
+    Q_INVOKABLE void setBlur(QQuickItem *item, bool blur);
 };
 
-- 
2.40.0

